### DISCUSSION TITLE
Severe Questions about ArrayVariable

### DISCUSSION THREAD

**Discussion by dingqiushi** [idaholab/moose#21817] [Q&A General] [answered]
The inheritance relationship changed from inheriting ` Kernel.C`  to inheriting `ArrayKernel.C`. The data type of the variables also changed from "`const VariableValue & _u`;" to "`const ArrayVariableValue & _u`;". There are severe questions about data types that I would like to ask.

1. How do I express the component of `_u` in the code, which brackets should I use, [] or ()?

2. I need to use the value of _u in another file, so assign the value of _u in the kernel file to another file by means of a pointer. My current compilation succeeds in writing the assignment  in the kernel file
```
for (size_t i=0; i<_count; ++i)
  _cfd_data._u[i] = _u[i];
```
Declare the new variable in another file
`RealEigenVector _u[10];`
I don't know if this makes sense, is there any connection between the data types `RealEigenVector` and` ArrayVariableValue?`

3. I found the data type RealEigenMatrix in MooseType.h. If I replace the Real type in the code when I have a single variable with the RealEigenVector type which is suitable for multiple variables, should I replace RealVectorValue with RealEigenMatrix, and how should a variable of type RealEigenMatrix represent its components?

4. For the previous RealTensorValue type, I did not find a matching type suitable for multiple variables. Is there any data type that can represent a variable with more than 2 dimensions? For example, I want to represent a variable with a dimension of 5 x 5 x 3 x 3, which can be seen as a large matrix nested within a small matrix, where each component of the large 5 x 5 matrix is a small 3 x 3 matrix, what data type is suitable for such a variable?

5. I also found the data types RealVectorArrayValue and RealTensorArrayValue in MooseType.h. They both seem to have a limit on the number of columns, with a dynamic number of rows and a Moose::dim number of columns. Can the number of columns for these two data types exceed 3?

6. Can any variable in the code directly represent the order of _u of a multivariate? Which self-contained function in moose can be used to get the norm directly for variables of type RealEigenVector?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Please read the array variable documentation, I feel it will answer a lot of your questions.
https://mooseframework.inl.gov/source/variables/ArrayMooseVariable.html

For question 4, array is a 1D array and is your only option for a multi-component variable with an arbitrary number (here 225) of components IF that is the non linear variable. 

If that object is NOT the non linear variable, then you may look at using a RankFourTensor instead

6) we define the .norm() class member attribute routine for a lot of classes. I cant recall if RealEigenVector is one of them. 

Guillaume

Guillaume
